lib/concurrent/atomic/atomic_markable_reference.rb

```diff
@@ -0,0 +1,152 @@
+      #
+      #   @return [Boolean] the current marked value
+      def mark
+        @Reference.get[1]
+      end
+      alias_method :marked?, :mark
+
+      # @!macro [attach] atomic_markable_reference_method_set
+      #
+      #   _Unconditionally_ sets to the given value of both the reference and
+      #   the mark.
+      #
+      #   @param [Object] new_val the new value
+      #   @param [Boolean] new_mark the new mark
+      #
+      #   @return [ImmutableArray] both the new value and the new mark
+      def set(new_val, new_mark)
+        ImmutableArray[new_val, new_mark].tap do |pair|
+          @Reference.set pair
+        end
+      end
+
+      # @!macro [attach] atomic_markable_reference_method_update
+      #
+      # Pass the current value and marked state to the given block, replacing it
+      # with the block's results. May retry if the value changes during the
+      # block's execution.
+      #
+      # @yield [Object] Calculate a new value and marked state for the atomic
+      #   reference using given (old) value and (old) marked
+      # @yieldparam [Object] old_val the starting value of the atomic reference
+      # @yieldparam [Boolean] old_mark the starting state of marked
+      #
+      # @return [ImmutableArray] the new value and new mark
+      def update
+        loop do
+          old_val, old_mark = value, marked?
+          new_val, new_mark = yield old_val, old_mark
+
+          if compare_and_set old_val, new_val, old_mark, new_mark
+            return ImmutableArray[new_val, new_mark]
+          end
+        end
+      end
+
+      # @!macro [attach] atomic_markable_reference_method_try_update
+      #
+      # Pass the current value to the given block, replacing it
+      # with the block's result. Raise an exception if the update
+      # fails.
+      #
+      # @yield [Object] Calculate a new value and marked state for the atomic
+      #   reference using given (old) value and (old) marked
+      # @yieldparam [Object] old_val the starting value of the atomic reference
+      # @yieldparam [Boolean] old_mark the starting state of marked
+      #
+      # @return [ImmutableArray] the new value and marked state
+      #
+      # @raise [Concurrent::ConcurrentUpdateError] if the update fails
+      def try_update
+        old_val, old_mark = value, marked?
+        new_val, new_mark = yield old_val, old_mark
+
+        unless compare_and_set old_val, new_val, old_mark, new_mark
+          fail ::Concurrent::ConcurrentUpdateError,
+               'AtomicMarkableReference: Update failed due to race condition.',
+               'Note: If you would like to guarantee an update, please use ' \
+               'the `AtomicMarkableReference#update` method.'
+        end
+
+        ImmutableArray[new_val, new_mark]
+      end
+
+      # Internal/private ImmutableArray for representing pairs
+      class ImmutableArray < Array
+        def self.new(*args)
+          super(*args).freeze
+        end
+      end
```

Comment:
The problem with `freeze` is `unfreeze` sort of defeats the purpose. I'm working on an immutable struct over [here](https://github.com/ruby-concurrency/concurrent-ruby/tree/immutable-struct). I'll do some benchmarking once it's further along so we can see how efficient it is in relation to an array. It may be the better choice.